pymodbus/client/async/asyncio/__init__.py

```diff
@@ -0,0 +1,777 @@
+
+    def protocol_made_connection(self, protocol):
+        """
+        Protocol notification of successful connection.
+        """
+        _logger.info('Protocol made connection.')
+        if not self.connected:
+            self.connected = True
+            self.protocol = protocol
+        else:
+            _logger.error('Factory protocol connect '
+                          'callback called while connected.')
+
+    def protocol_lost_connection(self, protocol):
+        """
+        Protocol notification of lost connection.
+        """
+        if self.connected:
+            _logger.info('Protocol lost connection.')
+            if protocol is not self.protocol:
+                _logger.error('Factory protocol callback '
+                              'called from unexpected protocol instance.')
+
+            self.connected = False
+            self.protocol = None
+            # if self.host:
+            #    asyncio.async(self._reconnect(), loop=self.loop)
+        else:
+            _logger.error('Factory protocol disconnect '
+                          'callback called while not connected.')
+
+
+class AsyncioModbusSerialClient(object):
+    """
+    Client to connect to modbus device over serial.
+    """
+    transport = None
+    framer = None
+
+    def __init__(self, port, protocol_class=None, framer=None,  loop=None):
+        """
+        Initializes Asyncio Modbus Serial Client
+        :param port: Port to connect
+        :param protocol_class: Protocol used to talk to modbus device.
+        :param framer: Framer to use
+        :param loop: Asyncio Event loop
+        """
+        #: Protocol used to talk to modbus device.
+        self.protocol_class = protocol_class or ModbusClientProtocol
+        #: Current protocol instance.
+        self.protocol = None
+        #: Event loop to use.
+        self.loop = loop or asyncio.get_event_loop()
+        self.port = port
+        self.framer = framer
+        self._connected = False
+
+    def stop(self):
+        """
+        Stops connection
+        :return:
+        """
+        if self._connected:
+            if self.protocol:
+                if self.protocol.transport:
+                    self.protocol.transport.close()
+
+    def _create_protocol(self):
+        """
+        Factory function to create initialized protocol instance.
+        """
+        from serial_asyncio import create_serial_connection
+
+        def factory():
+            return self.protocol_class(framer=self.framer)
+
+        cor = create_serial_connection(self.loop, factory, self.port)
+        return cor
+
+    @asyncio.coroutine
+    def connect(self):
+        """
+        Connect Async client
+        :return:
+        """
+        _logger.debug('Connecting.')
+        try:
+            yield from self.loop.create_connection(self._create_protocol)
+            _logger.info('Connected to %s:%s.' % (self.host, self.port))
```

Comment:
This class does not have `self.host` (and `host` don't make sense for serial connection).

Reply:
@pax0r thanks for taking the time to review , much appreciated. Lot of cleanup is still to be done with this branch. I am going to keep this PR for some more days for others to review as well and will take care of your comments.

Reply:
@dhoomakethu, I just started to work on project using asyncio and modbus communication and decided to give a try to your work. For now looks like asyncio part is working well, I'll come back if I will found any more problems.

Reply:
Thanks , just curious, are you dealing with the serial part or the TCP? Please raise an issue if you come across any.

Reply:
For now Serial only for communication with Fatek PLCs

Reply:
I recently did some work on the Serial Asyncio Modbus Client. The API of `serial_asyncio.create_serial_connection` is different from the socket API. Also, the serial kwargs (baudrate etc.) were missing in `AsyncioModbusSerialClient.__init__` Method. I didn't have time to test my changes because `serial_asyncio` does not yet work on windows. I will generate a pull request anyway for reference.